### DISCUSSION TITLE
Optional options

### DISCUSSION THREAD

**Discussion by lookfirst** [uber-go/fx#836] [Ideas]
I'd like to generate a bunch of options based on some conditional that is injected. This is kind of minimal pseudocode for what I'm trying to do. Help?

```golang
type Task interface {}

func NewFoo() Task {
}

func NewBar() Task {
}

type TaskGroup {
  fx.In
  Tasks []Task  `group:"tasks"`
}

var tasks = []interface{}{ NewBar, NewFoo }

// This is the part that is a real mess... and obviously won't work...
var Module = fx.Invoke(func (m machine.Machine) fx.Option {
   var options []fx.Option
   if m.IsFoo() {
	options = append(options, fx.Provide(fx.Annotate(t, fx.ResultTags(`group:"tasks"`))))
   }

   return fx.Options()
})

var TaskManagerModule = fx.Provide(func (tg TaskGroup) {
   // tg.Tasks is empty here, but should have the NewFoo Task...
})
```

**Comment by abhinav** [comment]
Hi there! This is something that has come up before in #698. The author of that post (and accompanying PRs) referred to the feature as "Late Provides". This isn't something that Fx supports right now, and we're reluctant to add support for it.

My comment on that issue (https://github.com/uber-go/fx/issues/698#issuecomment-961294249) explains our reasoning in some detail, but the gist of it is: a feature like this would make the Fx graph a lot more dynamic and harder to debug than it already is, and based on our experience running Fx at a large scale, we think a static graph is desirable for now.

We're open to reconsidering the position if there's consistent demand for it, but right now #698 reflects our position pretty well.
